File: android-app/app/src/main/java/com/securityradio/ptt/device/P25AmbeNative.kt
package com.securityradio.ptt.device

/**
 * JNI bridge to the bundled dvmvocoder's AMBE+2 half-rate mode (GPL-2.0; see
 * cpp/dvmvocoder/README.txt) — the P25 Phase 2 / DMR vocoder rate: 49 voice
 * bits @ 2450 bps in a 9-byte DMR-interleaved codeword per 20 ms frame.
 *
 * Shares `libsecurityradiovocoder.so` with [P25ImbeNative]; the 16 kHz ↔ 8 kHz
 * framing helpers live there too ([P25ImbeNative.Frames]).
 */
object P25AmbeNative {

    @Volatile
    private var loadedOk = false

    /** True once the JNI lib loaded and the AMBE codec pair allocated. */
    val isAvailable: Boolean get() = loadedOk

    fun tryLoadLibrary(): Boolean {
        if (loadedOk) return true
        return try {
            System.loadLibrary("securityradiovocoder")
            loadedOk = nativeInit()
            loadedOk
        } catch (_: Throwable) {
            loadedOk = false
            false
        }
    }

    /** PCM @ 8 kHz, **exactly** 160 samples → 9-byte AMBE codeword. */
    fun encodeFrame(samples8k160: ShortArray): ByteArray? {
        if (!loadedOk || samples8k160.size != 160) return null
        return nativeEncode(samples8k160)
    }

    fun decodeCodeword9(codeword9: ByteArray): ShortArray? {
        if (!loadedOk || codeword9.size != 9) return null
        return nativeDecode(codeword9)
    }

    @JvmStatic private external fun nativeInit(): Boolean
    @JvmStatic private external fun nativeEncode(samples8k160: ShortArray): ByteArray?
    @JvmStatic private external fun nativeDecode(codeword9: ByteArray): ShortArray?
}

File: ios-app/SafeTMobile/Audio/P25AmbeNative.swift
import Foundation

/// Swift wrapper around the bundled dvmvocoder's AMBE+2 half-rate mode — the
/// P25 Phase 2 / DMR vocoder rate: 49 voice bits @ 2450 bps in a 9-byte
/// DMR-interleaved codeword per 20 ms frame. Shares the native bridge with
/// `P25ImbeNative` (see Native/p25_vocoder_bridge.cpp); the 16 kHz ↔ 8 kHz
/// framing helpers live in `P25ImbeNative.Frames`.
enum P25AmbeNative {
    private static let lock = NSLock()
    private static var ready = false

    static var isAvailable: Bool {
        lock.lock()
        defer { lock.unlock() }
        return ready
    }

    @discardableResult
    static func initialize() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if ready { return true }
        ready = p25_ambe_init()
        return ready
    }

    static func encodeFrame(samples8k160: [Int16]) -> Data? {
        guard samples8k160.count == 160 else { return nil }
        lock.lock()
        defer { lock.unlock() }
        if !ready { return nil }
        var codeword = [UInt8](repeating: 0, count: 9)
        let ok = samples8k160.withUnsafeBufferPointer { samplesPtr in
            codeword.withUnsafeMutableBufferPointer { outPtr in
                guard let s = samplesPtr.baseAddress, let o = outPtr.baseAddress else { return false }
                return p25_ambe_encode(s, o)
            }
        }
        return ok ? Data(codeword) : nil
    }

    static func decodeCodeword9(_ codeword: Data) -> [Int16]? {
        guard codeword.count == 9 else { return nil }
        lock.lock()
        defer { lock.unlock() }
        if !ready { return nil }
        var samples = [Int16](repeating: 0, count: 160)
        let ok = codeword.withUnsafeBytes { raw in
            samples.withUnsafeMutableBufferPointer { outPtr in
                guard let c = raw.baseAddress?.assumingMemoryBound(to: UInt8.self),
                      let o = outPtr.baseAddress else { return false }
                return p25_ambe_decode(c, o)
            }
        }
        return ok ? samples : nil
    }
}
